libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
+		s.Overrides = append(s.Overrides, "options")
+	}
+
+	templText := `{{if .Comments}}{{with .Md.HeaderComment}}{{.}}
+
+{{end}}{{end}}{{range .NameServers -}}
+nameserver {{.}}
+{{end}}{{with .Search -}}
+search {{.}}
+{{end}}{{with .Options -}}
+options {{.}}
+{{end}}{{with .Other -}}
+{{.}}
+{{end}}{{if .Comments}}
+# Based on host file: '{{.Md.SourcePath}}'{{with .Md.Transform}} ({{.}}){{end}}
+{{if .Md.UsedDefaultNS -}}
+# Used default nameservers.
+{{end -}}
+{{with .ExtNameServers -}}
+# ExtServers: {{.}}
+{{end -}}
+# Overrides: {{.Overrides}}
+{{with .Md.NDotsFrom -}}
+# Option ndots from: {{.}}
+{{end -}}
+{{end -}}
+`
+	var buf bytes.Buffer
+	templ, err := template.New("summary").Parse(templText)
+	if err != nil {
+		return nil, errdefs.System(err)
+	}
+	if err := templ.Execute(&buf, s); err != nil {
+		return nil, errdefs.System(err)
+	}
+	return buf.Bytes(), nil
+}
+
+// WriteFile generates content and writes it to path. If hashPath is non-zero, it
+// also writes a file containing a hash of the content, to enable UserModified()
+// to determine whether the file has been modified.
+func (rc *ResolvConf) WriteFile(path, hashPath string, perm os.FileMode) error {
+	content, err := rc.Generate(true)
+	if err != nil {
+		return err
+	}
+
+	// Write the resolv.conf file - it's bind-mounted into the container, so can't
+	// move a temp file into place, just have to truncate and write it.
+	if err := os.WriteFile(path, content, perm); err != nil {
+		return errdefs.System(err)
+	}
+
+	// Write the hash file.
+	if hashPath != "" {
+		hashFile, err := ioutils.NewAtomicFileWriter(hashPath, perm)
+		if err != nil {
+			return errdefs.System(err)
+		}
+		defer hashFile.Close()
+
+		_, err = hashFile.Write(hashData(content))
+		if err != nil {
+			return errdefs.System(err)
+		}
+	}
+
+	return nil
+}
+
+// UserModified can be used to determine whether the resolv.conf file has been
+// modified since it was generated. It returns true if the files at rcPath and
+// rcHashPath both exist, and the hash file does not contain the hash of the
+// current rc file - else false.
+func UserModified(rcPath, rcHashPath string) bool {
+	currRC, err := os.ReadFile(rcPath)
+	if err != nil {
+		return !os.IsNotExist(err)
+	}
+
+	currRCHash, err := os.ReadFile(rcHashPath)
+	if err != nil {
+		return !os.IsNotExist(err)
+	}
+
+	return !bytes.Equal(hashData(currRC), currRCHash)
+}
+
+func readResolvConf(path string, reader io.Reader) (*ResolvConf, error) {
+	var rc ResolvConf
+	rc.md.SourcePath = path
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		rc.processLine(scanner.Text())
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, errdefs.System(err)
+	}
+	return &rc, nil
+}
+
+func (rc *ResolvConf) processLine(line string) {
+	fields := strings.Fields(line)
+
+	// Strip comments.
+	for i, s := range fields {
+		if s[0] == '#' || s[0] == ';' {
+			fields = fields[:i]
+			break
+		}
+	}
+	if len(fields) == 0 {
+		return
+	}
```

Comment:
Agreed - and that's how I wrote it initially (except it needs to deal with blank lines too). Then I found the unit tests that explicitly check for comments at the end of lines ...

https://github.com/moby/moby/blob/43ffb1ee9d5adeca444d9f5d061313ba617910a1/libnetwork/resolvconf/resolvconf_unix_test.go#L204-L231

So, this was to preserve existing behaviour - which seemed safest, but I should have added a comment.

It is wrong though - for example, resolvers treat "comments" on the end of an options line as lots of options. Most of the comment chatter will be ignored but, for example, `options # ndots:1` does set the `ndots` option.

I could fix it as part of this PR but, as far as possible, was trying to avoid making drive-by fixes like that.